src/util.ts

```diff
@@ -29,13 +29,9 @@
 import type { TomlDate } from './date.js'
 import { TomlError } from './error.js'
 
-export type TomlPrimitive =
-	| string
-	| number
-	| boolean
-	| TomlDate
-	| { [key: string]: TomlPrimitive }
-	| TomlPrimitive[]
+export type TomlPrimitive = string | number | boolean | TomlDate
+export type TomlTable = { [Key in string]: TomlValue }
+export type TomlValue = TomlPrimitive | TomlValue[] | TomlTable
```

Comment:
- There's no reason to use a mapped type here; it should just be a regular index type (or a Record)
- Style: I'd prefer for the recursion to happen last in TomlValue

Reply:
Record won't work, because typescript doesn't allow recursion through it.
I don't know the difference between index type and mapped type, I'll trust you with it.

In TomlValue's three possible values, only TomlPrimitive doesn't recurse back to TomlValue. I find it more straightforward to have the more direct version first, then increasing levels of complexity. If that's what you meant.